File: Go/Go_1000-/1403.go
// Go
// leetcode 1403
// https://leetcode.cn/problems/minimum-subsequence-in-non-increasing-order/

func minSubsequence(nums []int) []int {
	n := len(nums)
	tmp := make([]int, n)
	copy(tmp, nums)
	sort.Slice(tmp, func(i, j int) bool {
		return tmp[i] > tmp[j]
	})
	sum := 0
	for _, v := range nums {
		sum += v
	}
	cur, i := 0, 0
	for cur * 2 <= sum {
		cur += tmp[i]
		i++
	}

	return tmp[0 : i]
}

File: cpp/cpp_1000-/1403.cc
// C++
// leetcode 1403
// https://leetcode.cn/problems/minimum-subsequence-in-non-increasing-order/

class Solution {
public:
    vector<int> minSubsequence(vector<int>& nums) {
    	priority_queue<int> pq(nums.begin(), nums.end());

    	int sum = 0;
    	for(auto x : nums)
    		sum += x;

    	int cur = 0;
    	vector<int> res;
    	while (cur * 2 <= sum) {
    		cur += pq.top();
    		res.push_back(pq.top());
    		pq.pop();
    	}

    	return res;
    }
};